### PR TITLE
skt use marker offset 

### PR DESCRIPTION
**Pull request by dydkask**

### CODE REVIEW

**Review by jorgecoca — CHANGES_REQUESTED** [review]
Can you please give some information about why these changes are necessary? 

Based on the code submitted, I don't understand what's the goal of these changes

**Comment by dydkask** [comment]
> Can you please give some information about why these changes are necessary?
> 
> Based on the code submitted, I don't understand what's the goal of these changes

The code is designed to solve MOB-2702.
And to use offset in SKT-ATLAS, atlas modification is necessary.
If you have any questions, please tell me.

**Comment by jorgecoca** [comment]
Understood. I don't believe that implementation details of SKT should affect Atlas, and therefore, I don't believe this change should be reflected in this project.

I would encourage you to find a fix for your issue within the SKT project, without modifying Atlas, since Atlas has nothing to do with it.

If you agree with that, I would like to close this PR without merging it. Thanks!

**Comment by felangel** [comment]
I agree with @jorgecoca. SKT-specific functionality should not require changes in Atlas.

**Review by felangel — CHANGES_REQUESTED** [review]
see comments

**Comment by ksantiyanont** [comment]
@jorgecoca/@felangel I think Jaeyoung's explanation may not be complete. This change is to allow us to control how we anchor the marker icon, this is not specific to any map provider implementation. It is true that most of the marker icons have the tip at the center bottom, but for example the vehicle location icon usually have the location anchored at the center. There is no way for the plugin to tell how each icon image looks like. Having the offsets is the only way to keep everything generic.

**Comment by jorgecoca** [comment]
> @jorgecoca/@felangel I think Jaeyoung's explanation may not be complete. This change is to allow us to control how we anchor the marker icon, this is not specific to any map provider implementation. It is true that most of the marker icons have the tip at the center bottom, but for example the vehicle location icon usually have the location anchored at the center. There is no way for the plugin to tell how each icon image looks like. Having the offsets is the only way to keep everything generic.

I think this explanation makes a bit more sense, now I have a better understanding of the problem... however, I still think the code (well, the documentation) is fairly incomplete: as a user consuming this API, I would never imagine that those two variables are doing what they are doing, and the fact that they are already defaulted to 0.5 seems very confusing.

What are the units of the offset? How much impact does a 0.5? Why are we using `offsetX/Y` if a marker only has lat/lng?

I'd suggest giving a much more detailed explanation in the code documentation, and not giving an offset by default, since it will impact all the markers

**Comment by felangel** [comment]
I agree with @jorgecoca's comments and think we should potentially explore different alternatives to solve this problem.

Might be nicer to have an API where you tell the marker how to anchor itself:

```dart
/// {@template anchor}
/// Specifies the anchor to be at a particular point in the marker image.
///
/// The anchor specifies the point in the icon image that is anchored
/// to the marker's position on the Earth's surface.
///
/// The anchor point is specified in the continuous space [0.0, 1.0] x [0.0, 1.0],
/// where (0, 0) is the top-left corner of the image, and (1, 1) is the bottom-right corner.
/// The anchoring point in a W x H image is the nearest discrete grid point in a
/// (W + 1) x (H + 1) grid, obtained by scaling the then rounding.
/// For example, in a 4 x 2 image, the anchor point (0.7, 0.6) resolves
/// to the grid point at (3, 1).
/// {@endtemplate}
class Anchor {
  /// u-coordinate of the anchor, as a ratio of the image width (in the range [0, 1])
  final float u;
  /// v-coordinate of the anchor, as a ratio of the image height (in the range [0, 1])
  final float v;
  /// {@macro anchor}
  const Anchor({@required this.u, @required this.v});
}
```

```dart
  /// Anchor point of the marker
  final Anchor anchor;

  const Marker({
    @required this.id,
    @required this.position,
    this.onTap,
    this.icon,
    this.anchor,
    this.zIndex = 0.0,
  });
```

Thoughts?

**Comment by ksantiyanont** [comment]
@felangel I like what you proposed. Two things to mention is that (1) both anchor coordinates can be outside of the [0, 1] range, which indicate that the anchor point is outside of the icon image, and (2) we may want to mention what does it mean when the anchor parameter is not set. Normally, we should default this to (0.5, 0.5).

**Comment by felangel** [comment]
> @felangel I like what you proposed. Two things to mention is that (1) both anchor coordinates can be outside of the [0, 1] range, which indicate that the anchor point is outside of the icon image, and (2) we may want to mention what does it mean when the anchor parameter is not set. Normally, we should default this to (0.5, 0.5).

In what case would we want an anchor to be outside of the marker icon? That doesn't make sense to me because you shouldn't be able to anchor outside of the marker's width/height. I think defaulting to (0.5, 0.5) makes sense 👍

**Comment by ksantiyanont** [comment]
> > @felangel I like what you proposed. Two things to mention is that (1) both anchor coordinates can be outside of the [0, 1] range, which indicate that the anchor point is outside of the icon image, and (2) we may want to mention what does it mean when the anchor parameter is not set. Normally, we should default this to (0.5, 0.5).
> 
> In what case would we want an anchor to be outside of the marker icon? That doesn't make sense to me because you shouldn't be able to anchor outside of the marker's width/height. I think defaulting to (0.5, 0.5) makes sense 👍

@felangel I cannot think of a good use case for that either, but it adds flexibility if the app needs the marker to be off by a pixel or two.

**Comment by jorgecoca** [comment]
Do you think it makes sense to document this feature in the README as well?

File: packages/atlas/lib/src/marker.dart [hunk 1]
```diff
@@ -45,3 +48,14 @@ class Marker {
   int get hashCode =>
       id.hashCode ^ position.hashCode ^ icon.hashCode ^ zIndex.hashCode;
 }
+
```
Comment (felangel) [inline packages/atlas/lib/src/marker.dart]: I think you need to add the {@template anchor} and other docs from my comment here in order for the @macro to work

File: packages/atlas/test/marker_test.dart [hunk 2]
```diff
@@ -119,7 +119,31 @@ main() {
         onTap: () {},
         zIndex: 1.0,
       );
+
+      final anchor = new Anchor();
```
Comment (felangel) [inline packages/atlas/test/marker_test.dart]: Can you remove the new keyword please?

**Review by felangel — CHANGES_REQUESTED** [review]
See comments

File: packages/atlas/test/marker_test.dart (hunk 2 above)
Comment (dydkask) [reply packages/atlas/test/marker_test.dart]: i fix that

File: packages/atlas/lib/src/marker.dart (hunk 1 above)
Comment (dydkask) [reply packages/atlas/lib/src/marker.dart]: i fix that

**Comment by dydkask** [comment]
already merge

**Pull request closed without merging**